load_webp.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2021 Tobias Stoeckmann <tobias@stoeckmann.org>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <err.h>
+#include <limits.h>
+#include <pixman.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <webp/decode.h>
+
+#include "functions.h"
+
+
+pixman_image_t *
+load_webp(FILE *fp)
+{
+	pixman_image_t *img;
+	uint8_t *data, *data_buffer, *p;
+  uint32_t *pixel, *pixels;
+	size_t data_size;
+	unsigned int width, height, ok, len;
+
+  fseek(fp, 0, SEEK_END);
+  data_size = ftell(fp);
+  fseek(fp, 0, SEEK_SET);
```

Comment:
The code does not cover error conditions. The worst part would be ftell returning -1. Since data_size is an unsigned type, you would have all bits set to 1. Which leads to...

Reply:
The whole setup of fseek+ftell looks like fstat would be a better choice (but even then look out for off_t size_t differences in st_size).